omegaconf/dictconfig.py

```diff
@@ -107,24 +107,51 @@ def __init__(
             format_and_raise(node=None, key=None, value=None, cause=ex, msg=str(ex))
 
     def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "DictConfig":
-        res = DictConfig({})
-        for k, v in self.__dict__.items():
-            res.__dict__[k] = copy.deepcopy(v, memo=memo)
-        res._re_parent()
+        res = DictConfig(None)
+        res.__dict__["_metadata"] = copy.deepcopy(self.__dict__["_metadata"], memo=memo)
+        res.__dict__["_flags_cache"] = copy.deepcopy(
+            self.__dict__["_flags_cache"], memo=memo
+        )
+
+        src_content = self.__dict__["_content"]
+        if isinstance(src_content, dict):
+            content_copy = {}
+            for k, v in src_content.items():
+                try:
+                    old_parent = v.__dict__["_parent"]
+                    v.__dict__["_parent"] = None
+                    vc = copy.deepcopy(v, memo=memo)
+                    vc.__dict__["_parent"] = res
+                    content_copy[k] = vc
+                finally:
+                    v.__dict__["_parent"] = old_parent
+        else:
+            # None and strings can be assigned as is
+            content_copy = src_content
+
+        res.__dict__["_content"] = content_copy
```

Comment:
Not directly related to this line:

There's actually a functional change: now, the parent of a deep-copied DictConfig (and also ListConfig) is `None`. I believe it's ok (and better than dragging around the old parent in the copy), but:

1. There should be a test for this (I didn't see one added in the commit that introduced this new implementation)
2. Ideally this would be the case with the deepcopy of *any* node, not just containers

Reply:
Oh you're right I forgot about interpolations!

I think you're going to need to deepcopy the parent as well though, as otherwise it breaks the usual assumption that the (deep) copy is independent from the original object.

Reply:
That's the optimization.
deepcopying the parent while you are copying a recursive structure is extremely expensive.

I think this is a reasonable tradeoff.
You are welcome to try to re-introducing deepcopying of the parent (on top of my other diff) while matintaining the optimization wins.